### PR TITLE
feat: adding golint and a Makefile

### PR DESCRIPTION
**Pull request by mfreeman451**
1. Added a `Makefile` that includes tests for golint, ineffassign, gosec, gocov, etc.
2. Adding `coverage.yml` to ci for gocov
3. Added `golintci-lint.yml` to ci for golint
4. Updated a test so it would pass gosec check

### CODE REVIEW

File: .github/workflows/coverage.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          persist-credentials: false # otherwise, the token used is the GITHUB_TOKEN, instead of your personal access token.
+          fetch-depth: 0 # otherwise, there would be errors pushing refs to the destination repository.
+
+      - name: Setup go
+        uses: actions/setup-go@v2
+        with:
+          go-version: '1.20'
```
Comment (tmc) [inline .github/workflows/coverage.yml:22]: this is fine for now but with setup-go@v3 they support a "stable" alias which would lower toil of maintaining this over time.

Comment (tmc) [reply .github/workflows/coverage.yml:22]: also setup-go v4 does some automatic caching I think (we can add this later)

File: .github/workflows/golangci-lint.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+- name: Run golangci-lint
```
Comment (tmc) [inline .github/workflows/golangci-lint.yml:1]: I'm not familiar with workflows being this short in nature -> doesn't this need a top level name and then a job defined?

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,86 @@
+.PHONY: \
```
Comment (tmc) [inline Makefile:1]: I have a makefile coming in a recent other change that may impact this, FYI

File: go.work [hunk 4]
```diff
@@ -1,3 +1,4 @@
 go 1.20
```
Comment (tmc) [inline go.work:1]: as go.work files aren't intended to be shared I think we should pull this out of this PR

File: llms/openai/openaillm.go [hunk 5]
```diff
@@ -10,7 +10,7 @@ import (
 	"github.com/tmc/langchaingo/llms/openai/internal/openaiclient"
 )
 
-const tokenEnvVarName = "OPENAI_API_KEY"
+var tokenEnvVarName = os.Getenv("OPENAI_API_KEY")
```
Comment (tmc) [inline llms/openai/openaillm.go:13]: this seems to change the semantics here -> EnvVarName vs EnvVar value?

File: outputParsers/emptyParser.go [hunk 6]
```diff
@@ -2,7 +2,7 @@ package outputParsers
 
 import "github.com/tmc/langchaingo/prompts"
 
-//Output parser that does nothing
```
Comment (tmc) [inline outputParsers/emptyParser.go:5]: can we hold off on these unrelated changes?

File: llms/openai/openaillm.go (hunk 5 above)
Comment (mfreeman451) [reply llms/openai/openaillm.go:13]: I would have named it something else I guess, but I was just trying to make it pass a gosec check.

Comment (mfreeman451) [reply llms/openai/openaillm.go:13]: oh i see what you're saying, let me double check that

Comment (mfreeman451) [reply llms/openai/openaillm.go]: fixed

File: .github/workflows/golangci-lint.yml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      - uses: actions/setup-go@v4
+        with:
+          go-version: '1.19'
+          cache: false
+      - uses: actions/checkout@v3
+      - name: golangci-lint
+        uses: golangci/golangci-lint-action@v3
+        with:
+          # Optional: version of golangci-lint to use in form of v1.2 or v1.2.3 or `latest` to use the latest version
+          version: v1.29
```
Comment (nekomeowww) [inline .github/workflows/golangci-lint.yml:28]: Shouldn't we use the latest version?

```suggestion
          version: v1.52.2
```

File: .github/workflows/golangci-lint.yml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+          # Optional: if set to true then the all caching functionality will be complete disabled,
+          #           takes precedence over all other caching options.
+          # skip-cache: true
+
+          # Optional: if set to true then the action don't cache or restore ~/go/pkg.
+          # skip-pkg-cache: true
+
+          # Optional: if set to true then the action don't cache or restore ~/.cache/go-build.
+          # skip-build-cache: true
```
Comment (nekomeowww) [inline .github/workflows/golangci-lint.yml:47]: Should these comments be deleted? Developers could go to the repository of `golangci/golangci-lint-action` find out the parameters.

And there should be a new empty line for the last line.
```suggestion
          
```

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,6 @@
+coverage.txt
+go.work
+go.sum
```
Comment (nekomeowww) [inline .gitignore:3]: `go.sum` should never be ignored.

According to Golang's official documentation of [Modules](https://go.dev/blog/using-go-modules):

> The go command uses the go.sum file to ensure that future downloads of these modules retrieve the same bits as the first download, to ensure the modules your project depends on do not change unexpectedly, whether for malicious, accidental, or other reasons. **Both go.mod and go.sum should be checked into version control**.

```suggestion
```

File: .github/workflows/coverage.yml [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        uses: tj-actions/verify-changed-files@v12
+        id: verify-changed-files
+        with:
+          files: README.md
+
+      - name: Commit changes
+        if: steps.verify-changed-files.outputs.files_changed == 'true'
+        run: |
+          git config --local user.email "action@github.com"
+          git config --local user.name "GitHub Action"
```
Comment (nekomeowww) [inline .github/workflows/coverage.yml:51]: In order to get the bot avatar and info looks properly, should use the following email:

```suggestion
          git config user.name 'GitHub Action'
          git config user.email 'github-actions[bot]@users.noreply.github.com'
```

Read more about: https://github.com/actions/checkout/discussions/479

File: .github/workflows/coverage.yml [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    branches:
+      - main
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    name: Update coverage badge
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
```
Comment (nekomeowww) [inline .github/workflows/coverage.yml:14]: Should use `actions/checkout@v3` from this time period

```suggestion
        uses: actions/checkout@v3
```

File: examples/QA-example/go.mod [hunk 12]
```diff
@@ -1,3 +1,7 @@
 module langchaingo/examples/vectors-example
 
 go 1.20
+
+require github.com/tmc/langchaingo v0.0.0-20230417231738-0d76b082ba42
```
Comment (nekomeowww) [inline examples/QA-example/go.mod:5]: The `go.mod` shouldn't require itself. What happened here?

Comment (nekomeowww) [inline examples/QA-example/go.mod:5]: How about adding a `replace` command above to keep the dependent modules up-to-date when testing and building?

However, it is not traditional to have a standalone `example` folder in Golang. Golang has its own support of examples: https://go.dev/blog/examples. Such example functions can be directly executed on `go.pkg` documentation site. Should it be replaced by `ExampleXXX`, @tmc ?

File: .gitignore [hunk 13]
```diff
@@ -0,0 +1,6 @@
+coverage.txt
```
Comment (nekomeowww) [inline .gitignore:1]: Coverage output should ends with `.out`, what is this for specificly?

File: .github/workflows/golangci-lint.yml [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+  # Optional: allow read access to pull request. Use with `only-new-issues` option.
+  # pull-requests: read
+jobs:
+  golangci:
+    name: lint
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/setup-go@v4
+        with:
+          go-version: '1.19'
```
Comment (nekomeowww) [inline .github/workflows/golangci-lint.yml:21]:
```suggestion
          go-version: '1.20'
```

The `go.mod` file says it is 1.20 for this project.

Comment (nekomeowww) [reply .github/workflows/golangci-lint.yml:21]: Or you can specify the same value as the coverage one.

File: .github/workflows/coverage.yml [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+        with:
+          persist-credentials: false # otherwise, the token used is the GITHUB_TOKEN, instead of your personal access token.
+          fetch-depth: 0 # otherwise, there would be errors pushing refs to the destination repository.
+
+      - name: Setup go
+        uses: actions/setup-go@v2
+        with:
+          go-version: '1.20'
+
+      - uses: actions/cache@v2
```
Comment (nekomeowww) [inline .github/workflows/coverage.yml:24]: Should use actions/cache@v3 from this time period

```suggestion
      - uses: actions/cache@v3
```

File: .github/workflows/golangci-lint.yml [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+          cache: false
```
Comment (nekomeowww) [inline .github/workflows/golangci-lint.yml:22]: Why the `cache` option is disabled? Is there any concern about this?

File: .github/workflows/golangci-lint.yml [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+      - v*
+    branches:
+      - master
+      - main
+  pull_request:
+permissions:
+  contents: read
+  # Optional: allow read access to pull request. Use with `only-new-issues` option.
+  # pull-requests: read
+jobs:
```
Comment (nekomeowww) [inline .github/workflows/golangci-lint.yml:14]: Could we have the same spaces between the sections?

```suggestion
name: golangci-lint

on:
  push:
    tags:
      - v*
    branches:
      - master
      - main
  pull_request:
  
permissions:
  contents: read
  # Optional: allow read access to pull request. Use with `only-new-issues` option.
  # pull-requests: read
  
jobs:
```

File: .github/workflows/golangci-lint.yml [hunk 18]
```diff
@@ -0,0 +1,47 @@
+name: golangci-lint
```
Comment (nekomeowww) [inline .github/workflows/golangci-lint.yml:1]: I suggest changing the `name` to `Lint` and add one more job to scan vulnerability:

```yaml
  # TODO: Should be removed after golangci-lint integrated with vuln
  vulnerability:
    name: Vulnerabilities Checks
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3

      - uses: actions/setup-go@v3
        with:
          go-version: "^1.20.0"
          cache: true

      - uses: actions/cache@v3
        with:
          path: ~/go/pkg/mod
          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
          restore-keys: |
            ${{ runner.os }}-go-

      - name: Scan vulnerability
        run: |
          go install golang.org/x/vuln/cmd/govulncheck@latest
          govulncheck ./...
```

More about vuln: https://go.dev/blog/vuln

And `golangci-lint` is working on to add the integration support with [`vuln`](https://github.com/golangci/golangci-lint/issues/3094), you can track the progress [here](https://github.com/golangci/golangci-lint/pull/3199)

File: .github/workflows/coverage.yml [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+        with:
+          files: README.md
+
+      - name: Commit changes
+        if: steps.verify-changed-files.outputs.files_changed == 'true'
+        run: |
+          git config --local user.email "action@github.com"
+          git config --local user.name "GitHub Action"
+          git add README.md
+          git commit -m "chore: Updated coverage badge."
```
[the remark this answers is not in the document]
Comment (nekomeowww) [reply .github/workflows/coverage.yml:53]: Or how about we make two workflow definition:

- Pull Request into `main`, focusing on testing and report the developers
- Push into `main`, focusing on updating the test coverages

**Comment by nekomeowww** [comment]
There are some conflicts. Remember to get your branch updated 😉

**Comment by rdbell** [comment]
Looking forward to a more uniform codebase with conventions to follow. Let's get this in.

File: .github/workflows/coverage.yml [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+          git config user.name 'GitHub Action'
+          git add README.md
+          git commit -m "chore: Updated coverage badge."
+
+      - name: Push changes
+        if: steps.verify-changed-files.outputs.files_changed == 'true'
+        uses: ad-m/github-push-action@master
+        with:
+          github_token: ${{ github.token }}
+          branch: ${{ github.head_ref }}
```
Comment (nekomeowww) [inline .github/workflows/coverage.yml:60]: Every file should ends with a empty new line.
```suggestion
          branch: ${{ github.head_ref }}
          
```

You can ask IDE/editors perform such actions automatically, if you are using Visual Studio Code, you could add one line that enable this feature easily:

```json
{
  "files.insertFinalNewline": true,
}
```

**Comment by tmc** [comment]
I believe this is obviated by https://github.com/tmc/langchaingo/pull/42
